File: backend/funzycode-backend/src/main/java/com/funzycode/backend/service/LevelService.java
package com.funzycode.backend.service;

import com.funzycode.backend.dto.CompleteLevelRequest;
import com.funzycode.backend.dto.LevelProgressResponse;
import com.funzycode.backend.dto.SaveCodeRequest;

import java.util.List;

public interface LevelService {
    List<LevelProgressResponse> getLevels();
    LevelProgressResponse getLevel(int levelNumber);
    LevelProgressResponse startLevel(int levelNumber);
    LevelProgressResponse saveCode(int levelNumber, SaveCodeRequest request);
    LevelProgressResponse completeLevel(int levelNumber, CompleteLevelRequest request);
    LevelProgressResponse buyHint(int levelNumber);
    LevelProgressResponse autoFix(int levelNumber);
    LevelProgressResponse extraTime(int levelNumber);
}

File: frontend/services/levelService.js
// services/levelService.js

import apiClient, { unwrapApiResponse } from "./apiClient";

export const levelService = {
  async getLevels() {
    const response = await apiClient.get("/levels");
    return unwrapApiResponse(response);
  },

  async getLevel(levelNumber) {
    const response = await apiClient.get(`/levels/${levelNumber}`);
    return unwrapApiResponse(response);
  },

  async startLevel(levelNumber) {
    const response = await apiClient.post(`/levels/${levelNumber}/start`);
    return unwrapApiResponse(response);
  },

  async saveCode(levelNumber, payload) {
    const response = await apiClient.post(`/levels/${levelNumber}/save-code`, payload);
    return unwrapApiResponse(response);
  },

  async complete(levelNumber, payload) {
    const response = await apiClient.post(`/levels/${levelNumber}/complete`, payload);
    return unwrapApiResponse(response);
  },

  async buyHint(levelNumber) {
    const response = await apiClient.post(`/levels/${levelNumber}/buy-hint`);
    return unwrapApiResponse(response);
  },

  async autoFix(levelNumber) {
    const response = await apiClient.post(`/levels/${levelNumber}/auto-fix`);
    return unwrapApiResponse(response);
  },

  async extraTime(levelNumber) {
    const response = await apiClient.post(`/levels/${levelNumber}/extra-time`);
    return unwrapApiResponse(response);
  },
};

export default levelService;
